Sources/SwiftProductGenerator.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  SwiftProductGenerator.swift
+//
+//
+//  Created by Øyvind Grimnes on 09/09/15.
+//
+//
+
+
+/**
+A cartesian product generator
+
+:discussion:    Generates the cartesian product of collections, or repetitions
+                of a collection. The rightmost element is advanced every
+                iteration which ensures that if the input is sorted, the
+                output will be sorted as well.
+*/
+public struct Product<T, C: Collection>: IteratorProtocol, Sequence
+    where C.Iterator.Element == T
+{
+
+    /// Private variable to manage where to get next element in each pool
+    private var indices : [C.Index]
+    /// Pool containing all the collections to be combined
+    private var pools   : [C]
+    /// Terminate the generation on completion
+    private var done    : Bool = false
+
+
+    /**
+    Initiates the generator with an array of collections to be combined.
+
+    :param:        collections    array containing the collections to be
+                                  combined
+    */
+    public init(_ collections: [C]) {
+        self.pools   = collections
+        self.indices = collections.map{ $0.startIndex }
+        self.done    = pools.reduce(true, { $0 && $1.count == 0 })
```

Comment:
The `reduce` approach isn't only compact, it is also much more efficient, as it doesn't require the creation of an array and only relies on a boolean accumulator. My goal was to write something similar to a `any` statement in Python.

To be honest, I personally don't think the filter approach is easier to understand. I remember taking some time to actually realise you only tested all arrays for emptiness. Unless you're strongly against it, I think it'd be better not to come back to that approach.